In [103]:
import pandas as pd 

df = pd.read_csv("cust_data.csv")

#print(df.columns)

# df.describe()
df.info()
df["TotalCharges"].info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

In [104]:
print(df["Churn"].value_counts())

Churn
No     5174
Yes    1869
Name: count, dtype: int64


In [105]:
df["TotalCharges"].describe()
print(df["TotalCharges"].value_counts())

TotalCharges
20.2      11
          11
19.75      9
19.9       8
20.05      8
          ..
1990.5     1
7362.9     1
346.45     1
306.6      1
6844.5     1
Name: count, Length: 6531, dtype: int64


In [106]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")
df = df.dropna()

In [107]:
#print(df[ms.isnull()])
print(df.head())

   customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract Pape

In [108]:
df["Churn"] = df["Churn"].map({"Yes":1, "No":0})
cm = df.groupby("Contract")["Churn"].mean()
print(cm)

Contract
Month-to-month    0.427097
One year          0.112772
Two year          0.028487
Name: Churn, dtype: float64


In [109]:
tm = df.groupby("tenure")["Churn"].mean()
print(tm)

tenure
1     0.619902
2     0.516807
3     0.470000
4     0.471591
5     0.481203
        ...   
68    0.090000
69    0.084211
70    0.092437
71    0.035294
72    0.016575
Name: Churn, Length: 72, dtype: float64


In [110]:
mm = df.groupby("Churn")["MonthlyCharges"].mean()
print(mm)

Churn
0    61.307408
1    74.441332
Name: MonthlyCharges, dtype: float64


In [111]:
print(df.groupby("TechSupport")["Churn"].mean())

TechSupport
No                     0.416475
No internet service    0.074342
Yes                    0.151961
Name: Churn, dtype: float64


In [112]:
from sklearn.model_selection import train_test_split 
X = df.columns
X = df.drop(["customerID","Churn"], axis=1)   
y = df["Churn"]
X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=42,test_size=0.2,stratify=y)

In [113]:
df.info()

<class 'pandas.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   str    
 1   gender            7032 non-null   str    
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   str    
 4   Dependents        7032 non-null   str    
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   str    
 7   MultipleLines     7032 non-null   str    
 8   InternetService   7032 non-null   str    
 9   OnlineSecurity    7032 non-null   str    
 10  OnlineBackup      7032 non-null   str    
 11  DeviceProtection  7032 non-null   str    
 12  TechSupport       7032 non-null   str    
 13  StreamingTV       7032 non-null   str    
 14  StreamingMovies   7032 non-null   str    
 15  Contract          7032 non-null   str    
 16  PaperlessBilling  7032 non-null   str    
 17  PaymentMeth

In [114]:
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder()
categorical_columns = X_train.select_dtypes(include="object").columns
encoded_train = encoder.fit_transform(X_train[categorical_columns])
numerical_columns = X_train.select_dtypes(include =["int64","float64"]).columns

C:\Users\omkar\AppData\Local\Temp\ipykernel_7568\3235119527.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = X_train.select_dtypes(include="object").columns


In [115]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_features = scaler.fit_transform(X_train[numerical_columns])
print(scaled_features)


[[-0.43931886  1.3218163   0.98155578  1.6599004 ]
 [-0.43931886 -0.26741023 -0.97154551 -0.56225219]
 [-0.43931886  1.4440645   0.83706615  1.75610395]
 ...
 [-0.43931886  0.14008375  0.92674937  0.47350714]
 [-0.43931886 -0.9194006   0.02991714 -0.721544  ]
 [ 2.27625101 -1.28614519  0.32221802 -0.9787973 ]]


In [116]:
from sklearn.compose import ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numerical_columns),
        ("cat", OneHotEncoder(), categorical_columns)
    ]
)
X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)
print(X_train_processed.shape)
print(X_test_processed.shape)


(5625, 45)
(1407, 45)


In [117]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()

model.fit(X_train_processed, y_train)

y_pred = model.predict(X_test_processed)

In [118]:

from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred)

print(cm)

[[917 116]
 [160 214]]


In [119]:
from sklearn.metrics import accuracy_score, precision_score, recall_score

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))

Accuracy: 0.8038379530916845
Precision: 0.6484848484848484
Recall: 0.5721925133689839


In [120]:
from sklearn.tree import DecisionTreeClassifier
tree = DecisionTreeClassifier(max_depth=6)
tree.fit(X_train_processed, y_train)
train_score = tree.score(X_train_processed, y_train)
test_score = tree.score(X_test_processed, y_test)

print("Train:", train_score)
print("Test:", test_score)

Train: 0.8126222222222222
Test: 0.7903340440653873


In [121]:
tree_pred = tree.predict(X_test_processed)
print("Accuracy:", accuracy_score(y_test,tree_pred))
print("Precision:", precision_score(y_test,tree_pred))
print("Recall:", recall_score(y_test, tree_pred))

Accuracy: 0.7903340440653873
Precision: 0.616519174041298
Recall: 0.5588235294117647


# Customer Churn Prediction - Model Comparison

## Logistic Regression

- Accuracy: 80.4%
- Precision: 64.8%
- Recall: 57.2%

### Observations
- Strong baseline model.
- Performed better than the Decision Tree on all evaluated metrics.
- Generalized well to unseen data.

---

## Decision Tree

### Default Tree

- Train Accuracy: 99.9%
- Test Accuracy: 73.1%

### Observation
- Severe overfitting.
- Memorized training data but generalized poorly.

### Tuned Tree (max_depth=6)

- Accuracy: 79.0%
- Precision: 61.7%
- Recall: 55.9%

### Observation
- Reducing depth improved generalization.
- Still underperformed Logistic Regression.

---

## Conclusion

Logistic Regression is currently the best model.

Reasons:
- Higher Accuracy
- Higher Precision
- Higher Recall
- Better generalization

Decision Trees were able to model interactions between features, but the overfitting cost outweighed the benefits on this dataset.

In [122]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(max_depth=12, n_estimators=200)
rf.fit(X_train_processed, y_train)
print("Train:", rf.score(X_train_processed, y_train))
print("Test:", rf.score(X_test_processed, y_test))

Train: 0.9431111111111111
Test: 0.7882018479033405


In [123]:
rf_pred = rf.predict(X_test_processed)
print("Accuracy:", accuracy_score(y_test, rf_pred))
print("Precision:", precision_score(y_test, rf_pred))
print("Recall:", recall_score(y_test, rf_pred))

Accuracy: 0.7882018479033405
Precision: 0.6194968553459119
Recall: 0.5267379679144385


In [124]:
y_prob = model.predict_proba(X_test_processed)[:,1]
y_pred_03 = (y_prob >= 0.2).astype(int)
print("Accuracy:", accuracy_score(y_test, y_pred_03))
print("Precision:", precision_score(y_test, y_pred_03))
print("Recall:", recall_score(y_test, y_pred_03))


Accuracy: 0.6915422885572139
Precision: 0.45774647887323944
Recall: 0.8689839572192514


In [125]:
importance_df = pd.DataFrame({
    "Feature": preprocessor.get_feature_names_out(),
    "Importance": rf.feature_importances_
})
importance_df = importance_df.sort_values(
    by="Importance",
    ascending=False
)

print(importance_df.head(10))

                                Feature  Importance
1                           num__tenure    0.138384
3                     num__TotalCharges    0.135634
2                   num__MonthlyCharges    0.109211
36         cat__Contract_Month-to-month    0.077268
27                  cat__TechSupport_No    0.038525
43  cat__PaymentMethod_Electronic check    0.035059
18               cat__OnlineSecurity_No    0.033324
16     cat__InternetService_Fiber optic    0.032455
38               cat__Contract_Two year    0.023858
15             cat__InternetService_DSL    0.017069


In [126]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(
    model,
    X_train_processed,
    y_train,
    cv=5,
    scoring="recall"
)

print(scores)
print("Average Recall:", scores.mean())

[0.53846154 0.55852843 0.5451505  0.51839465 0.57525084]
Average Recall: 0.5471571906354515


In [128]:
from sklearn.model_selection import GridSearchCV
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [8, 10, 12, 15]
}

grid = GridSearchCV(
    RandomForestClassifier(random_state=42),
    param_grid=param_grid,
    cv=5,
    scoring='recall',
    n_jobs=-1
)
grid.fit(X_train_processed, y_train)
print(grid.best_params_)
print(grid.best_score_)

{'max_depth': 12, 'n_estimators': 100}
0.5210702341137123


In [129]:
best_rf = grid.best_estimator_

rf_pred = best_rf.predict(X_test_processed)

print("Accuracy:", accuracy_score(y_test, rf_pred))
print("Precision:", precision_score(y_test, rf_pred))
print("Recall:", recall_score(y_test, rf_pred))

Accuracy: 0.7910447761194029
Precision: 0.6351351351351351
Recall: 0.5026737967914439
